File: back/src/controllers/notificationController.js
import mongoose from 'mongoose';
import Notification from '../models/Notification.js';

export const getNotifications = async (req, res) => {
  try {
    const { targetUser } = req.query;
    const filter = {};
    if (targetUser) {
      filter.$or = [{ targetUser }, { targetUser: 'all' }];
    }
    const list = await Notification.find(filter).sort({ timestamp: -1 }).limit(100).lean();
    res.json(list);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
};

export const markAsRead = async (req, res) => {
  try {
    const { id } = req.params;
    await Notification.findOneAndUpdate(
      getIdFilter(id),
      { $set: { read: true } }
    );
    res.json({ success: true });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
};

export const markAllAsRead = async (req, res) => {
  try {
    const { targetUser } = req.body;
    const filter = targetUser ? { $or: [{ targetUser }, { targetUser: 'all' }] } : {};
    await Notification.updateMany(filter, { $set: { read: true } });
    res.json({ success: true });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
};

export const getSyncMessages = async (req, res) => {
  try {
    const { targetUser } = req.query;
    const filter = targetUser ? { $or: [{ targetUser }, { targetUser: 'all' }] } : {};
    const list = await Notification.find(filter).lean();
    res.json(list);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
};

const getIdFilter = (id) => {
  if (!id) return { id: 'invalid_id_placeholder' };
  const is24Hex = mongoose.Types.ObjectId.isValid(id) && String(new mongoose.Types.ObjectId(id)) === id;
  if (is24Hex) {
    return { _id: id };
  }
  return { id };
};

File: back/Controllers/NotificationController.cs
using System;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using Eftekad.Backend.Data;
using Eftekad.Backend.Models;
using System.Collections.Generic;
using System.Linq;

namespace Eftekad.Backend.Controllers
{
    [ApiController]
    [Route("api/notifications")]
    public class NotificationController : ControllerBase
    {
        private readonly EftekadDbContext _context;

        public NotificationController(EftekadDbContext context)
        {
            _context = context;
        }

        [HttpGet]
        public async Task<IActionResult> GetNotifications([FromQuery] string? targetUser)
        {
            try
            {
                var query = _context.Notifications.AsQueryable();

                if (!string.IsNullOrEmpty(targetUser))
                {
                    query = query.Where(n => n.TargetUser == targetUser || n.TargetUser == "all");
                }

                var list = await query
                    .OrderByDescending(n => n.Timestamp)
                    .Take(100)
                    .ToListAsync();

                return Ok(list);
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { error = ex.Message });
            }
        }

        [HttpPost("{id}/read")]
        public async Task<IActionResult> MarkAsRead(string id)
        {
            try
            {
                var notification = await _context.Notifications.FirstOrDefaultAsync(n => n.Id == id);
                if (notification != null)
                {
                    notification.Read = true;
                    notification.UpdatedAt = DateTime.UtcNow;
                    await _context.SaveChangesAsync();
                }
                return Ok(new { success = true });
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { error = ex.Message });
            }
        }

        public class MarkAllPayload
        {
            public string? TargetUser { get; set; }
        }

        [HttpPost("read-all")]
        public async Task<IActionResult> MarkAllAsRead([FromBody] MarkAllPayload payload)
        {
            try
            {
                var query = _context.Notifications.AsQueryable();

                if (!string.IsNullOrEmpty(payload.TargetUser))
                {
                    query = query.Where(n => n.TargetUser == payload.TargetUser || n.TargetUser == "all");
                }

                var list = await query.ToListAsync();
                foreach (var n in list)
                {
                    n.Read = true;
                    n.UpdatedAt = DateTime.UtcNow;
                }

                await _context.SaveChangesAsync();
                return Ok(new { success = true });
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { error = ex.Message });
            }
        }

        [HttpGet("sync-messages")]
        public async Task<IActionResult> GetSyncMessages([FromQuery] string? targetUser)
        {
            try
            {
                var query = _context.Notifications.AsQueryable();

                if (!string.IsNullOrEmpty(targetUser))
                {
                    query = query.Where(n => n.TargetUser == targetUser || n.TargetUser == "all");
                }

                var list = await query.ToListAsync();
                return Ok(list);
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { error = ex.Message });
            }
        }
    }
}
